File: shcloj2/code/project.clj
;;
;;Excerpted from "Programming Clojure, Second Edition",
;;published by The Pragmatic Bookshelf.
;;Copyrights apply to this code. It may not be used to create training material, 
;;courses, books, articles, and the like. Contact us if you are in doubt.
;;We make no guarantees that this code is fit for any purpose. 
;;Visit http://www.pragmaticprogrammer.com/titles/shcloj2 for more book information.
;;
(defproject programming-clojure "1.3.0"
  :description "Examples from Programming Clojure"
  :aot [examples.tasklist]
  :dependencies [[org.clojure/clojure "1.3.0"]
                 [org.clojure/java.jdbc "0.0.6"]
                 [org.clojure/test.generative "0.1.2"]
                 [ring "0.3.11" :exclusions [org.clojure/clojure
                                             org.clojure/clojure-contrib]]
                 [compojure "0.6.5"]
                 [hiccup "0.3.6"]
                 [jline "0.9.94"]
                 [hsqldb "1.8.0.10"]])

File: warv/code/artflow/presenters/app/models/project.rb
#---
# Excerpted from "The Rails View",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material, 
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose. 
# Visit http://www.pragmaticprogrammer.com/titles/warv for more book information.
#---
class Project < ActiveRecord::Base
  belongs_to :campaign
  belongs_to :client
  has_many :creations, dependent: :destroy
  has_many :assignments, class_name: 'ProjectAssignment'
  has_many :designers, through: :assignments
  
  validates :name, presence: true
  validates :client_id, presence: true

  before_validation :copy_client_id, on: :create
  
  scope :active, where(active: true)

  def self.total_hours
    all.map(&:total_hours).sum
  end
  
  def total_hours
    creations.sum(:hours)
  end
  
  private

  def copy_client_id
    write_attribute(:client_id, campaign.client_id)
  end

end
